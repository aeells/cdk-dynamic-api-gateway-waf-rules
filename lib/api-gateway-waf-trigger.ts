import {Construct} from 'constructs';

import * as triggers from 'aws-cdk-lib/triggers';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as path from 'path';

export interface ApiGatewayWafTriggerProps
{
    readonly restApiId: string;
    readonly stageName: string;
}

export class ApiGatewayWafTrigger extends Construct
{
    readonly triggerFunction: triggers.TriggerFunction;

    constructor(scope: Construct, id: string, properties: ApiGatewayWafTriggerProps)
    {
        super(scope, id);

        this.triggerFunction = new triggers.TriggerFunction(this, 'ApiGatewayOpenApiPutFn', {
            runtime: lambda.Runtime.NODEJS_16_X,
            handler: 'index.handler',
            code: lambda.Code.fromAsset(path.join(__dirname, 'waf')),
            environment: {
              REST_API_ID: properties.restApiId,
              STAGE_NAME: properties.stageName
            },
            initialPolicy: [
                new iam.PolicyStatement({
                    effect: iam.Effect.ALLOW,
                    actions: ['apigateway:GET', 'ssm:*'],
                    resources: ['*']
                })
            ]
        })
    }
}