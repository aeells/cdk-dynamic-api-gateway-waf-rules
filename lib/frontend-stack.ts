import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import * as waf from 'aws-cdk-lib/aws-wafv2';
import * as ssm from 'aws-cdk-lib/aws-ssm';
import { Fn } from 'aws-cdk-lib';

export interface WafRule
{
    name: string;
    rule: waf.CfnWebACL.RuleProperty;
}

export class FrontEndStack extends cdk.Stack
{
    readonly openApiPaths: string[];

    constructor(scope: Construct, id: string, props?: cdk.StackProps)
    {
        super(scope, id, props);

        this.openApiPaths = Fn.split(',',
            ssm.StringParameter.fromStringParameterAttributes(this, 'ERSchedulerSSMOpenApiPaths', {
                parameterName: 'openapi-paths',
            }).stringValue
        );

        this.createWafAcl();
    }

    private createWafAcl() {
        const wafRules: WafRule[] = [
            ...this.getApiGatewayPathAllowRules(10), // 10+ Allow
            this.getApiGatewayPathBlockRule(20), // 20 Block
        ];

        return new waf.CfnWebACL(this, 'WafV2', {
            defaultAction: { allow: {} },
            visibilityConfig: {
                cloudWatchMetricsEnabled: false,
                metricName: 'cloudwatch-metrics-disabled-for-example-code',
                sampledRequestsEnabled: false,
            },
            scope: 'REGIONAL',
            rules: wafRules.map((wafRule) => wafRule.rule),
        });
    }

    // todo aeells - it might be more efficient to create a single rule with multiple statements?
    private getApiGatewayPathAllowRules(priority: number): WafRule[] {
        const rules: WafRule[] = [];
        for (let i = 0; i < this.openApiPaths.length; i++) {
            const apiPath: string = Fn.select(i, this.openApiPaths);
            rules.push(this.createApiGatewayRegexAllowRule(apiPath, priority++));
        }

        return rules;
    }

    private createApiGatewayRegexAllowRule(path: string, priority: number): WafRule {
        return {
            name: `ApiGatewayPathAllows${priority}`,
            rule: {
                name: `ApiGatewayPathAllows${priority}`,
                priority: priority,
                action: {
                    allow: {},
                },
                visibilityConfig: {
                    sampledRequestsEnabled: true,
                    cloudWatchMetricsEnabled: true,
                    metricName: 'ApiGatewayPathAllows',
                },
                statement: {
                    regexMatchStatement: {
                        fieldToMatch: {
                            uriPath: {},
                        },
                        regexString: path,
                        textTransformations: [
                            {
                                type: 'NONE',
                                priority: 0,
                            },
                        ],
                    },
                },
            },
        };
    }

    private getApiGatewayPathBlockRule(priority: number): WafRule {
        return {
            name: 'ApiGatewayPathBlocks',
            rule: {
                name: 'ApiGatewayPathBlocks',
                priority: priority,
                action: {
                    block: {},
                },
                visibilityConfig: {
                    sampledRequestsEnabled: true,
                    cloudWatchMetricsEnabled: true,
                    metricName: 'ApiGatewayPathBlocks',
                },
                statement: {
                    byteMatchStatement: {
                        fieldToMatch: {
                            uriPath: {},
                        },
                        positionalConstraint: 'STARTS_WITH',
                        searchString: '/api/',
                        textTransformations: [
                            {
                                type: 'NONE',
                                priority: 0,
                            },
                        ],
                    },
                },
            },
        };
    }
}
