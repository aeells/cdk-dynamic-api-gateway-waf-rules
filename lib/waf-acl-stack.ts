import {Construct} from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as waf from 'aws-cdk-lib/aws-wafv2';
import * as ssm from 'aws-cdk-lib/aws-ssm';

export interface WafRule
{
    name: string;
    rule: waf.CfnWebACL.RuleProperty;
}

export class WafAclStack extends cdk.Stack
{
    private openApiPaths: string[];

    constructor(scope: Construct, id: string, props?: cdk.StackProps)
    {
        super(scope, id, props);

        // retrieve the REST API endpoint paths from SSM
        this.retrieveRestApiEndpointPaths();

        // create WAF Web ACL rules locked down to those endpoints
        this.createWafAcl();
    }

    private retrieveRestApiEndpointPaths()
    {
        this.openApiPaths = cdk.Fn.split(',',
            ssm.StringParameter.fromStringParameterAttributes(this, 'SSMOpenApiPaths', {
                parameterName: 'api-paths',
            }).stringValue,
            // todo aeells - array 'assumedLength' is a restriction of the CDK and is required here to iterate properly
            // restriction documented in full in the Fn#split api https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-split.html

            // if your API is under active development and regularly  changing in size it would
            // be fairly simple to add a second 'array length' parameter to the Parameter Store
            2
        );
    }

    private createWafAcl() {
        const wafRules: WafRule[] = [
            ...this.getApiGatewayPathAllowRules(10), // 10+ Allow
            this.getApiGatewayPathBlockRule(20), // 20 Block
        ];

        return new waf.CfnWebACL(this, 'WafV2', {
            defaultAction: { allow: {} },
            visibilityConfig: {
                cloudWatchMetricsEnabled: false,
                metricName: 'cloudwatch-metrics-disabled-for-example-code',
                sampledRequestsEnabled: false,
            },
            scope: 'REGIONAL',
            rules: wafRules.map((wafRule) => wafRule.rule),
        });
    }

    private getApiGatewayPathAllowRules(priority: number): WafRule[] {
        const rules: WafRule[] = [];
        for (let i = 0; i < this.openApiPaths.length; i++) {
            const apiPath: string = cdk.Fn.select(i, this.openApiPaths);
            rules.push(this.createApiGatewayRegexAllowRule(apiPath, priority++));
        }

        return rules;
    }

    private createApiGatewayRegexAllowRule(path: string, priority: number): WafRule {
        return {
            name: `ApiGatewayPathAllows${priority}`,
            rule: {
                name: `ApiGatewayPathAllows${priority}`,
                priority: priority,
                action: {
                    allow: {},
                },
                visibilityConfig: {
                    sampledRequestsEnabled: true,
                    cloudWatchMetricsEnabled: true,
                    metricName: 'ApiGatewayPathAllows',
                },
                statement: {
                    regexMatchStatement: {
                        fieldToMatch: {
                            uriPath: {},
                        },
                        regexString: path,
                        textTransformations: [
                            {
                                type: 'NONE',
                                priority: 0,
                            },
                        ],
                    },
                },
            },
        };
    }

    private getApiGatewayPathBlockRule(priority: number): WafRule {
        return {
            name: 'ApiGatewayPathBlocks',
            rule: {
                name: 'ApiGatewayPathBlocks',
                priority: priority,
                action: {
                    block: {},
                },
                visibilityConfig: {
                    sampledRequestsEnabled: true,
                    cloudWatchMetricsEnabled: true,
                    metricName: 'ApiGatewayPathBlocks',
                },
                statement: {
                    byteMatchStatement: {
                        fieldToMatch: {
                            uriPath: {},
                        },
                        positionalConstraint: 'STARTS_WITH',
                        searchString: '/',
                        textTransformations: [
                            {
                                type: 'NONE',
                                priority: 0,
                            },
                        ],
                    },
                },
            },
        };
    }
}
