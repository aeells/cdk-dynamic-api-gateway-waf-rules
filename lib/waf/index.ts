import * as AWS from 'aws-sdk';

export const handler = async (
    event:  any
) => {
    const apiExport : any = await new AWS.APIGateway().getExport({
        exportType: 'oas30',
        restApiId: `${process.env.REST_API_ID}`,
        stageName: `${process.env.STAGE_NAME}`,
        accepts: 'application/json'
    }).promise();

    const apiPaths: string[] = [];
    for (const [key] of Object.entries(JSON.parse(apiExport.body.toString()).paths))
    {
        apiPaths.push(key.replace('{bookId}', '[A-Za-z0-9]+-[A-za-z0-9]+'));
    }

    await new AWS.SSM().putParameter({
        Name: 'api-paths',
        Value: apiPaths.toString(),
        Overwrite: true,
        Type: 'String'
    }).promise();
};

