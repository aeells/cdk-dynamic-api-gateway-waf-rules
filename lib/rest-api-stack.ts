import {Construct} from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import { ApiGatewayWafTrigger } from "./api-gateway-waf-trigger";

export class RestApiStack extends cdk.Stack
{
    private restApi: apigateway.RestApi;
    
    constructor(scope: Construct, id: string, props?: cdk.StackProps)
    {
        super(scope, id, props);

        // in the real world, our api endpoint resources are loaded from an external package from another development team
        this.createDummyRestApi();

        // there is no cdk api to list/get RestApi resources so using a lambda and sdk
        this.createApiGatewayWafTrigger();
    }

    private createDummyRestApi()
    {
        this.restApi = new apigateway.RestApi(this, 'ApiGatewayRestApi', {
            description: 'example api gateway',
        });

        const books = this.restApi.root.addResource('books');
        books.addMethod('GET');
        books.addMethod('POST');

        const book = books.addResource('{bookId}');
        book.addMethod('GET');
        book.addMethod('DELETE');

        // ...etc...
    }

    private createApiGatewayWafTrigger()
    {
        const apiGatewayWafTrigger = new ApiGatewayWafTrigger(this, 'ApiGatewayWafTrigger', {
            restApiId: this.restApi.restApiId,
            stageName: 'prod',
        });

        apiGatewayWafTrigger.triggerFunction.executeAfter(this.restApi);
    }
}
