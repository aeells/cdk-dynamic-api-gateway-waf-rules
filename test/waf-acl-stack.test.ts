import { Template } from 'aws-cdk-lib/assertions';
import { App } from 'aws-cdk-lib';
import {RestApiStack} from '../lib/rest-api-stack';
import {WafAclStack} from "../lib/waf-acl-stack";

describe('WAF ACL stack test', () => {
    const app = new App();

    const wafAclStack = new WafAclStack(app, 'FrontEndStack', {
        env: { account: process.env.CDK_DEFAULT_ACCOUNT, region: process.env.CDK_DEFAULT_REGION },
    });

    test('a WAF Web ACL exists with WAF "Allow" rules from deployed API endpoints and "Block" rules for all other endpoints as per source-controlled snapshot', () => {
        expect(Template.fromStack(wafAclStack)).toMatchSnapshot();
    });
});
