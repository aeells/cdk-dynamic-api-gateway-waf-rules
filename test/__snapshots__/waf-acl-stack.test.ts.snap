// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`REST API stack test a WAF Web ACL exists with WAF allow rules from API endpoints as per source-controlled snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "ERSchedulerSSMOpenApiPathsParameter": {
      "Default": "openapi-paths",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "WafV2": {
      "Properties": {
        "DefaultAction": {
          "Allow": {},
        },
        "Rules": [
          {
            "Action": {
              "Allow": {},
            },
            "Name": "ApiGatewayPathAllows10",
            "Priority": 10,
            "Statement": {
              "RegexMatchStatement": {
                "FieldToMatch": {
                  "UriPath": {},
                },
                "RegexString": {
                  "Fn::Select": [
                    0,
                    {
                      "Fn::Split": [
                        ",",
                        {
                          "Ref": "ERSchedulerSSMOpenApiPathsParameter",
                        },
                      ],
                    },
                  ],
                },
                "TextTransformations": [
                  {
                    "Priority": 0,
                    "Type": "NONE",
                  },
                ],
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "ApiGatewayPathAllows",
              "SampledRequestsEnabled": true,
            },
          },
          {
            "Action": {
              "Block": {},
            },
            "Name": "ApiGatewayPathBlocks",
            "Priority": 20,
            "Statement": {
              "ByteMatchStatement": {
                "FieldToMatch": {
                  "UriPath": {},
                },
                "PositionalConstraint": "STARTS_WITH",
                "SearchString": "/api/",
                "TextTransformations": [
                  {
                    "Priority": 0,
                    "Type": "NONE",
                  },
                ],
              },
            },
            "VisibilityConfig": {
              "CloudWatchMetricsEnabled": true,
              "MetricName": "ApiGatewayPathBlocks",
              "SampledRequestsEnabled": true,
            },
          },
        ],
        "Scope": "REGIONAL",
        "VisibilityConfig": {
          "CloudWatchMetricsEnabled": false,
          "MetricName": "cloudwatch-metrics-disabled-for-example-code",
          "SampledRequestsEnabled": false,
        },
      },
      "Type": "AWS::WAFv2::WebACL",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
